### ISSUE TITLE
ID Attribute Accessibility

### ISSUE THREAD

**Issue by irmerk** [accordproject/cicero-ui#56]
# WCAG 2.0 Level A [4.1.1](https://www.w3.org/TR/UNDERSTANDING-WCAG20/ensure-compat-parses.html) (Parsing)

## Tasks
- All `id` attributes should only be used once.

## Related Issues: 
- Issue [#41](https://github.com/accordproject/template-studio-v2/issues/41) in Template Studio v2

**Comment by irmerk** [comment]
### Hacktoberfest:
This issue can possibly provide for multiple pull requests.

**Comment by sameshl** [comment]
@irmerk You can assign me this issue. I would love to work on this. Could you just elaborate a bit on what exactly needs to be done here?

Thanks!

**Comment by irmerk** [comment]
Sure @sameshl, I think I would be looking into guides online on how to be compliant with WCAG 2.1 Level A 4.1.1.

**Comment by monika** [comment]
I don't mean to step on any toes, but I discovered one duplicated svg `id` in this repo while debugging https://github.com/accordproject/template-studio-v2/issues/41

I can submit a PR for that if @sameshl hasn't already.

The other duplicate svg `id`s are in `markdown-editor` which I can take care of if you'd like?

**Comment by irmerk** [comment]
@sameshl up to you. If they timeout for a few days, @monika it'll be all yours.

> The other duplicate svg `id`s are in `markdown-editor` which I can take care of if you'd like?

Sure, I think working on that issue in `markdown-editor` would be great.

**Comment by sameshl** [comment]
@monika You can go ahead. I haven't started working on this issue as on now.

**Comment by monika** [comment]
I'm new to [running unit tests](https://github.com/accordproject/template-studio-v2/blob/master/DEVELOPERS.md#unit-tests) but it looks like one of the tests uses the duplicated `id` value we need to change on `<ClauseComponent />` so it fails. How should I proceed?

<img width="1042" alt="Screen Shot 2019-10-05 at 8 45 57 PM" src="https://user-images.githubusercontent.com/316762/66263984-c8774c80-e7b1-11e9-8b0a-f5a9ede17412.png">

**Comment by irmerk** [comment]
Where is `icon-trash` coming from? This looks like it's the `svg` for the trash icon, where the `id` is set here:
- https://github.com/accordproject/cicero-ui/blob/master/src/icons/trash.js#L6
- https://github.com/accordproject/cicero-ui/blob/master/src/components/ClauseComponent.js#L16

**Comment by monika** [comment]
Yeah! So because these SVGs are being embedded into the markup as opposed to linked to, their markup is being included in the DOM - including their `id` attributes of which we're seeing duplicates, which is what's causing the validation issue.

For this SVG trash icon, I changed the duplicated `id` from `Page-1` to a unique `id` of `icon-trash`. We could name it anything, really - `toolbar-icon-trash` if we want something even more unique/specific. We can even remove it entirely if the `id` attributes on these icons are not being used to target them. 

We can also change the icons over to use HTML `class` attributes instead if we're using CSS to target any of their inner paths for styling. Since the `id` attribute values are repeated, using an HTML `class` would leave those values intact and it would validate, although we wouldn't be able to reference these SVGs via an `id` anymore.

You can see more of the duplicated `id`s in the [Markdown Editor repo's icons](https://github.com/accordproject/markdown-editor/tree/master/src/icons): 

https://github.com/accordproject/markdown-editor/blob/master/src/icons/OL.js#L15

**Comment by irmerk** [comment]
> Yeah! So because these SVGs are being embedded into the markup as opposed to linked to, their markup is being included in the DOM - including their `id` attributes of which we're seeing duplicates, which is what's causing the validation issue.

You mean they're being duplicated when more than one `ClauseComponent` exists on the page, correct?

> For this SVG trash icon, I changed the duplicated `id` from `Page-1` to a unique `id` of `icon-trash`. We could name it anything, really - `toolbar-icon-trash` if we want something even more unique/specific. 

If the answer to my previous question is yes, then this would need to be a `uuid`?

> We can even remove it entirely if the `id` attributes on these icons are not being used to target them.

Ah, I don't believe we are targeting them via somewhere else. I was under the impression that we needed them.... But we may actually not. @DianaLease thoughts?

> We can also change the icons over to use HTML `class` attributes instead if we're using CSS to target any of their inner paths for styling. Since the `id` attribute values are repeated, using an HTML `class` would leave those values intact and it would validate, although we wouldn't be able to reference these SVGs via an `id` anymore. It looks as though the fill colors are being set in-line though, so I'm guessing this won't be an issue but you might know better.

Hmm, so it seems it may be better to just remove the `id` from the `ClauseComponent` icons altogether. Looking for feedback on this.

**Comment by monika** [comment]
> > Yeah! So because these SVGs are being embedded into the markup as opposed to linked to, their markup is being included in the DOM - including their `id` attributes of which we're seeing duplicates, which is what's causing the validation issue.
> 
> You mean they're being duplicated when more than one `ClauseComponent` exists on the page, correct?

If `ClauseComponent` can exist more then once in the app, yeah we'll run into the duplication problem, yeah. 

_Also_ the `id` duplication issue is happening because the SVG icons in the text styling toolbar are also using `id="Page-1`, and several of their nested groups and paths are using `id="Path"`, or `id="Rectangle"`. Because those `id` values exist in more than one of the SVG icons, we're getting the duplication error.

All of the SVG icons, from this specific icon in this repo, to the icons in the Markdown Editor repo, are sharing `id` the same values and we can either change those values to be specific to each icon, or remove them entirely if they aren't being used.

**Comment by irmerk** [comment]
Yes this is definitely helpful! Thanks for all your contribution on this.

This seems like a mistake when creating the SVG icons for the `FormattingToolbar`. Sounds like we _could_ change them to be unique, but because we don't necessarily use them we should probably just remove them.

**Comment by monika** [comment]
Happy to help! It looks like the unit test for the `ClauseComponent` includes two `id` properties for the trash can icon though, so the unit test fails if I update or remove the `id`s:

https://github.com/accordproject/cicero-ui/blob/master/src/components/__snapshots__/index.test.js.snap#L27

https://github.com/accordproject/cicero-ui/blob/master/src/components/__snapshots__/index.test.js.snap#L38

Should the test be updated? I haven't worked with unit testing much so I'm not sure how to proceed on that front.

**Comment by irmerk** [comment]
Yes they should. When you've updated the code and run the test, near the bottom should be a command to update the snapshots (`npm test -- -u` or something similar).

**Comment by irmerk** [comment]
Thank you for all your investigating on these issues @monika. Feel like this closes the issue?

**Comment by monika** [comment]
You're welcome, thanks for organizing such a great way to help via Hacktoberfest!

As far as the audit via [axe](https://www.deque.com/axe/) is showing me, there are no duplicated `id` attributes that it can see. I think this specific issue is closed, yeah!

It might be worth adding a linter or unit test to check `id` attribute usage on components going forward, or run an audit on a semi-regular basis to just check that this is still the case.
